In [186]:
!nbmerge data_collection.ipynb sql_analysis.ipynb > presentation.ipynb

# Exploration & Analysis

In [143]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [178]:
%sql mysql://admin:sql_2020@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project

'Connected: admin@sql_project'

In [145]:
%%sql
SELECT *
FROM Customers
LIMIT 3;

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
3 rows affected.


[('0002-ORFBO', '34.827662', '34.827662, -118.999073', '-118.999073', 'Female', 'No', 'Yes', 'No', 9, 'Yes', 'No', 'DSL', 'No', 'Yes', 'No', 'Yes', 'Yes', 'No', 'One year', 'Yes', 'Mailed check', 65.6, 593.3, 'No', 0, 65, 2205, ''),
 ('0003-MKNFE', '34.162515', '34.162515, -118.203869', '-118.203869', 'Male', 'No', 'No', 'No', 9, 'Yes', 'Yes', 'DSL', 'No', 'No', 'No', 'No', 'No', 'Yes', 'Month-to-month', 'No', 'Mailed check', 59.9, 542.4, 'No', 0, 66, 5414, ''),
 ('0004-TLHLJ', '33.645672', '33.645672, -117.922613', '-117.922613', 'Male', 'No', 'No', 'No', 4, 'Yes', 'No', 'Fiber optic', 'No', 'No', 'Yes', 'No', 'No', 'No', 'Month-to-month', 'Yes', 'Electronic check', 73.9, 280.85, 'Yes', 1, 71, 4479, 'Price too high')]

# Exploration

## What is our churn rate?

In [146]:
%%sql
SELECT
    `Churn Value`,
    COUNT(`Churn Value`) AS `Count of Churn Value`,
    COUNT(`Churn Value`) / (SELECT COUNT(`Churn Value`) FROM Customers) AS PercentChurn
FROM Customers
GROUP BY `Churn Value`;

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
2 rows affected.


[(0, 5174, Decimal('0.7346')), (1, 1869, Decimal('0.2654'))]

Our churn rate is 26.54%, meaning that ~25% of our customers cancel their service

## Why did customers churn?

In [177]:
%%sql
SELECT `Churn Reason`, COUNT(`Churn Reason`) AS `Count of churn reason`, `Churn Value`
FROM Customers
GROUP BY `Churn Reason`
HAVING `Churn Value` = 1
ORDER BY COUNT(`Churn Reason`) DESC;

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
20 rows affected.


[('Attitude of support person', 192, 1),
 ('Competitor offered higher download speeds', 189, 1),
 ('Competitor offered more data', 162, 1),
 ("Don't know", 154, 1),
 ('Competitor made better offer', 140, 1),
 ('Attitude of service provider', 135, 1),
 ('Competitor had better devices', 130, 1),
 ('Network reliability', 103, 1),
 ('Product dissatisfaction', 102, 1),
 ('Price too high', 98, 1),
 ('Service dissatisfaction', 89, 1),
 ('Lack of self-service on Website', 88, 1),
 ('Extra data charges', 57, 1),
 ('Moved', 53, 1),
 ('Long distance charges', 44, 1),
 ('Limited range of services', 44, 1),
 ('Lack of affordable download/upload speed', 44, 1),
 ('Poor expertise of phone support', 20, 1),
 ('Poor expertise of online support', 19, 1),
 ('Deceased', 6, 1)]

The most common reasons for a customer cancelling thier subscription come from the attitudes of service and support providers and competitors offering better services. 

## What is our monthly lost income from churned customers?

If we were to keep our customers with the same monthly charges as before they churned, how much would we be making monthly?

In [148]:
%%sql
SELECT `Churn Value`, 
SUM(`Monthly Charges`) AS `Total Monthly Income`, 
SUM(`Monthly Charges`) / (SELECT SUM(`Monthly Charges`) FROM Customers) AS `Percentage of Total Monthly Income`
FROM Customers
GROUP BY `Churn Value`

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
2 rows affected.


[(0, 316985.75013160706, 0.6949664845142681),
 (1, 139130.8500137329, 0.3050335154857318)]

It seems about 1/3 of our historical monthly income has been lost due to churn: about $140,000

#### While during exploration I was looking to answer the question "Why are customers churning?", the Churn Reason feature sufficiently answers this question. However, there are many variables that can help answer questions about how to target new customers

# Analysis

# Big Question: What new customers should we target?

In order to make the most of our marketing budget, we need to know what demographics to focus on. My approach to this is to find, of our current customers, which demographics have a significantly lower _churn rate_ and/or significantly higher _monthly payments_. This will help differentiate which demographics will be better customers, and therefore will be worth reaching via marketing.

In the interest of simplification and not repeating myself, I will make a VIEW of all my tables already joined. This will be used in my later queries.

SQL Features Used: VIEW, JOIN

In [149]:
%%sql
CREATE OR REPLACE VIEW customerLocationFinances AS
    SELECT Customers.*, Locations.Zip, Finances.*
    FROM Customers
    JOIN Locations
        ON Customers.Latlong = Locations.LatLong
    JOIN Finances
        ON Locations.County = Finances.County

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
0 rows affected.


[]

## What kinds of customers will give us the highest average monthly income?

By finding which demographics of current customers provide the most revenue, we can predict which demographics would provide more revenue if they were to become customers. Therefore, we could market to those demographics to raise our revenue. 

This SQL query is an example query that is used in a stored procedure, as it will be run for all demographics of interest (ex. Gender, Has Dependents, Zip Code, etc). In this example, we will see the difference in average monthly income between senior citizens and non senior citizens. 

SQL Features Used: GROUP BY, Subquery, Window Function

In [ ]:
%%sql
DELIMITER $$
DROP PROCEDURE IF EXISTS getDifferencesInMonthlyRevenue$$
CREATE PROCEDURE getDifferencesInMonthlyRevenue (demographic VARCHAR(255))
	BEGIN
		SET @statement := CONCAT(
			'SELECT
			RANK() OVER(ORDER BY AVG(`Monthly Charges`) - (SELECT AVG(`Monthly Charges`) FROM customerLocationFinances) DESC) AS Ranking,
			`', demographic, '`,
			COUNT(`', demographic, '`) AS `Number of Customers`,
			ROUND(SUM(`Monthly Charges`), 2) AS `Total Monthly Revenue`,
			ROUND(AVG(`Monthly Charges`), 2) AS `Average Monthly Revenue`,
			ROUND(AVG(`Monthly Charges`) - (SELECT AVG(`Monthly Charges`) FROM customerLocationFinances),2) AS `Difference from statewide average`
			FROM customerLocationFinances
			GROUP BY `', demographic, '`');
		PREPARE differenceStatement FROM @statement;
		EXECUTE differenceStatement;
	END$$
DELIMITER ;

This is the stored procedure that allows the code above to be re-run on selected columns (demographics)

In [182]:
%%sql
CALL getDifferencesInMonthlyRevenue('Senior Citizen');

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
2 rows affected.


[(1, 'Yes', 1111, 88791.45, 79.92, 15.23),
 (2, 'No', 5787, 357460.85, 61.77, -2.92)]

Based on these results, it seems that senior citizens are a good choice for targeting, as they provide on average $15.23 more per month than the general average. 

We can also use this query to find locations to focus our marketing. By finding which counties pay a higher than average monthly revenue, we can determine where customers are more likely to get a bigger plan.

In [183]:
%%sql
CALL getDifferencesInMonthlyRevenue('County');

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
58 rows affected.


[(1, 'Tuolumne County', 44, 3214.65, 73.06, 8.37),
 (2, 'Napa County', 32, 2261.25, 70.66, 5.97),
 (3, 'Glenn County', 28, 1970.0, 70.36, 5.66),
 (4, 'El Dorado County', 80, 5624.9, 70.31, 5.62),
 (5, 'Mariposa County', 32, 2249.55, 70.3, 5.61),
 (6, 'Contra Costa County', 136, 9508.2, 69.91, 5.22),
 (7, 'Inyo County', 36, 2483.65, 68.99, 4.3),
 (8, 'Placer County', 120, 8222.85, 68.52, 3.83),
 (9, 'Tehama County', 52, 3537.75, 68.03, 3.34),
 (10, 'Ventura County', 102, 6933.0, 67.97, 3.28),
 (11, 'Riverside County', 298, 20161.85, 67.66, 2.96),
 (12, 'Stanislaus County', 96, 6490.45, 67.61, 2.92),
 (13, 'Sutter County', 36, 2431.6, 67.54, 2.85),
 (14, 'Siskiyou County', 88, 5923.3, 67.31, 2.62),
 (15, 'San Diego County', 455, 30574.45, 67.2, 2.5),
 (16, 'Mendocino County', 96, 6450.25, 67.19, 2.5),
 (17, 'Fresno County', 196, 13072.8, 66.7, 2.0),
 (18, 'San Bernardino County', 325, 21554.8, 66.32, 1.63),
 (19, 'Santa Clara County', 224, 14791.05, 66.03, 1.34),
 (20, 'San Benito County', 16, 1051.2, 65.7, 1.01),
 (21, 'San Joaquin County', 100, 6548.75, 65.49, 0.79),
 (22, 'San Mateo County', 112, 7327.6, 65.43, 0.73),
 (23, 'Sonoma County', 140, 9141.1, 65.29, 0.6),
 (24, 'Amador County', 36, 2343.6, 65.1, 0.41),
 (25, 'Butte County', 64, 4160.95, 65.01, 0.32),
 (26, 'Orange County', 346, 22484.25, 64.98, 0.29),
 (27, 'Lake County', 52, 3377.95, 64.96, 0.27),
 (28, 'Sacramento County', 196, 12728.2, 64.94, 0.25),
 (29, 'Lassen County', 56, 3629.85, 64.82, 0.13),
 (30, 'San Francisco County', 104, 6722.8, 64.64, -0.05),
 (31, 'Modoc County', 36, 2310.9, 64.19, -0.5),
 (32, 'Santa Barbara County', 92, 5903.5, 64.17, -0.52),
 (33, 'Los Angeles County', 1342, 85925.55, 64.03, -0.67),
 (34, 'Tulare County', 136, 8646.75, 63.58, -1.11),
 (35, 'Monterey County', 112, 7110.45, 63.49, -1.21),
 (36, 'San Luis Obispo County', 72, 4557.3, 63.3, -1.4),
 (37, 'Sierra County', 32, 2023.7, 63.24, -1.45),
 (38, 'Kern County', 164, 10363.55, 63.19, -1.5),
 (39, 'Del Norte County', 16, 1008.25, 63.02, -1.68),
 (40, 'Kings County', 24, 1509.55, 62.9, -1.8),
 (41, 'Colusa County', 28, 1747.85, 62.42, -2.27),
 (42, 'Yuba County', 48, 2991.9, 62.33, -2.36),
 (43, 'Merced County', 68, 4223.0, 62.1, -2.59),
 (44, 'Imperial County', 70, 4335.7, 61.94, -2.75),
 (45, 'Alameda County', 184, 11343.8, 61.65, -3.04),
 (46, 'Yolo County', 48, 2934.45, 61.13, -3.56),
 (47, 'Solano County', 52, 3172.85, 61.02, -3.68),
 (48, 'Alpine County', 4, 242.95, 60.74, -3.96),
 (49, 'Shasta County', 104, 6288.4, 60.47, -4.23),
 (50, 'Trinity County', 52, 3120.75, 60.01, -4.68),
 (51, 'Humboldt County', 136, 8155.5, 59.97, -4.73),
 (52, 'Calaveras County', 72, 4216.5, 58.56, -6.13),
 (53, 'Madera County', 44, 2562.6, 58.24, -6.45),
 (54, 'Marin County', 108, 6285.9, 58.2, -6.49),
 (55, 'Santa Cruz County', 64, 3511.55, 54.87, -9.83),
 (56, 'Plumas County', 44, 2319.0, 52.7, -11.99),
 (57, 'Mono County', 20, 1038.35, 51.92, -12.78),
 (58, 'Nevada County', 28, 1429.15, 51.04, -13.65)]

It seems Tuolumne county has the highest average monthly revenue at 73.06; 8.37 higher than the statewide average. I would recommend focusing on marketing in the top 5-10 counties from this query, as other customers in the county are likely to get similar plans

## What kinds of customers have a low churn rate?

To ensure any customers we generate will stay customers and not "churn" or cancel thier subscription, we want to market demographics with historically low churn rates. Thankfully, our dataset includes customers that have churned (~25% of our customers), along with a churn score which predicts the likelyhood of a customer churning (from 0-100)

This is the stored procedure that allows the code above to be re-run on selected columns (demographics)

In [ ]:
%%sql
DELIMITER $$
DROP PROCEDURE IF EXISTS getDifferencesInChurnRate$$
CREATE PROCEDURE getDifferencesInChurnRate (demographic VARCHAR(255))
	BEGIN
		SET @statement := CONCAT(
			'SELECT 
				RANK() OVER(ORDER BY AVG(`Churn Score`) - (SELECT AVG(`Churn Score`) FROM customerLocationFinances)) AS Ranking,
				`', demographic, '`,
				COUNT(`', demographic, '`) AS `Number of Customers`,
				SUM(`Churn Value`) AS `Count of Customers Churned`,
				ROUND(((SUM(`Churn Value`)/ COUNT(`', demographic, '`)))*100,1) AS `Percentage Churned`,
				ROUND(AVG(`Churn Score`), 2) AS `Average Churn Score`,
				ROUND(AVG(`Churn Score`) - (SELECT AVG(`Churn Score`) FROM customerLocationFinances),2) AS `Difference from statewide average`
			FROM customerLocationFinances
			GROUP BY `', demographic, '`');
		PREPARE differenceStatement FROM @statement;
		EXECUTE differenceStatement;
	END$$
DELIMITER ;

Much like the queries above, this query compares demographics' average churn scores to the statewide average, as well as providing information such as number of customers in each demographic, how many have churned, and the percentage of each demographic that have already churned. 

Understanding which demographics have lower average churn scores mean they are less likely to cancel their subscription, and are therefore more valuable customers

SQL Features Used: GROUP BY, Subquery, Window Function

In [181]:
%%sql
CALL getDifferencesInChurnRate('Senior Citizen');

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
2 rows affected.


[(1, 'No', 5787, Decimal('1365'), Decimal('23.6'), Decimal('57.75'), Decimal('-0.94')),
 (2, 'Yes', 1111, Decimal('459'), Decimal('41.3'), Decimal('63.57'), Decimal('4.88'))]

It seems senior citizens have a higher average churn score than usual, meaning they are more likely to cancel their subscription. This conflicts with our earlier assessment to target senior citizens as they provide higher monthly revenue than others. I believe this difference in churn score is negligable compared to the higher averagy monthly revenue, and therefore we should still market to the group. 

In [180]:
%%sql
CALL getDifferencesInChurnRate('County');

 * mysql://admin:***@lmu-sql.c7wywgkwglrl.us-west-2.rds.amazonaws.com/sql_project
58 rows affected.


[(1, 'Mono County', 20, Decimal('1'), Decimal('5.0'), Decimal('41.60'), Decimal('-17.09')),
 (2, 'Alpine County', 4, Decimal('0'), Decimal('0.0'), Decimal('43.50'), Decimal('-15.19')),
 (3, 'Mariposa County', 32, Decimal('5'), Decimal('15.6'), Decimal('52.78'), Decimal('-5.91')),
 (4, 'Lassen County', 56, Decimal('10'), Decimal('17.9'), Decimal('54.50'), Decimal('-4.19')),
 (5, 'Solano County', 52, Decimal('12'), Decimal('23.1'), Decimal('54.87'), Decimal('-3.82')),
 (6, 'Lake County', 52, Decimal('10'), Decimal('19.2'), Decimal('55.13'), Decimal('-3.56')),
 (7, 'Sutter County', 36, Decimal('9'), Decimal('25.0'), Decimal('55.17'), Decimal('-3.52')),
 (8, 'Mendocino County', 96, Decimal('21'), Decimal('21.9'), Decimal('55.18'), Decimal('-3.51')),
 (9, 'Merced County', 68, Decimal('18'), Decimal('26.5'), Decimal('55.28'), Decimal('-3.41')),
 (10, 'Humboldt County', 136, Decimal('32'), Decimal('23.5'), Decimal('55.74'), Decimal('-2.95')),
 (11, 'Tehama County', 52, Decimal('13'), Decimal('25.0'), Decimal('56.77'), Decimal('-1.92')),
 (12, 'Modoc County', 36, Decimal('11'), Decimal('30.6'), Decimal('56.78'), Decimal('-1.91')),
 (13, 'Calaveras County', 72, Decimal('18'), Decimal('25.0'), Decimal('57.26'), Decimal('-1.43')),
 (14, 'Madera County', 44, Decimal('7'), Decimal('15.9'), Decimal('57.36'), Decimal('-1.33')),
 (15, 'Placer County', 120, Decimal('30'), Decimal('25.0'), Decimal('57.58'), Decimal('-1.12')),
 (16, 'San Mateo County', 112, Decimal('28'), Decimal('25.0'), Decimal('57.60'), Decimal('-1.09')),
 (17, 'Sonoma County', 140, Decimal('33'), Decimal('23.6'), Decimal('57.61'), Decimal('-1.08')),
 (18, 'San Luis Obispo County', 72, Decimal('14'), Decimal('19.4'), Decimal('57.82'), Decimal('-0.87')),
 (19, 'Butte County', 64, Decimal('12'), Decimal('18.8'), Decimal('57.83'), Decimal('-0.86')),
 (20, 'Kern County', 164, Decimal('41'), Decimal('25.0'), Decimal('57.91'), Decimal('-0.78')),
 (21, 'Orange County', 346, Decimal('83'), Decimal('24.0'), Decimal('57.94'), Decimal('-0.75')),
 (22, 'Tuolumne County', 44, Decimal('13'), Decimal('29.5'), Decimal('58.02'), Decimal('-0.67')),
 (23, 'San Bernardino County', 325, Decimal('83'), Decimal('25.5'), Decimal('58.05'), Decimal('-0.64')),
 (24, 'Yolo County', 48, Decimal('14'), Decimal('29.2'), Decimal('58.06'), Decimal('-0.63')),
 (25, 'Santa Clara County', 224, Decimal('56'), Decimal('25.0'), Decimal('58.23'), Decimal('-0.46')),
 (26, 'Riverside County', 298, Decimal('77'), Decimal('25.8'), Decimal('58.27'), Decimal('-0.43')),
 (27, 'San Benito County', 16, Decimal('3'), Decimal('18.8'), Decimal('58.31'), Decimal('-0.38')),
 (28, 'Shasta County', 104, Decimal('31'), Decimal('29.8'), Decimal('58.65'), Decimal('-0.04')),
 (29, 'Los Angeles County', 1342, Decimal('358'), Decimal('26.7'), Decimal('58.66'), Decimal('-0.03')),
 (30, 'El Dorado County', 80, Decimal('23'), Decimal('28.8'), Decimal('58.70'), Decimal('0.01')),
 (31, 'Plumas County', 44, Decimal('12'), Decimal('27.3'), Decimal('58.91'), Decimal('0.22')),
 (32, 'Siskiyou County', 88, Decimal('20'), Decimal('22.7'), Decimal('58.94'), Decimal('0.25')),
 (33, 'Fresno County', 196, Decimal('51'), Decimal('26.0'), Decimal('59.02'), Decimal('0.33')),
 (34, 'Sacramento County', 196, Decimal('57'), Decimal('29.1'), Decimal('59.12'), Decimal('0.43')),
 (35, 'Sierra County', 32, Decimal('6'), Decimal('18.8'), Decimal('59.31'), Decimal('0.62')),
 (36, 'Santa Cruz County', 64, Decimal('16'), Decimal('25.0'), Decimal('59.38'), Decimal('0.68')),
 (37, 'Napa County', 32, Decimal('11'), Decimal('34.4'), Decimal('59.41'), Decimal('0.72')),
 (38, 'Amador County', 36, Decimal('9'), Decimal('25.0'), Decimal('59.42'), Decimal('0.73')),
 (39, 'Alameda County', 184, Decimal('50'), Decimal('27.2'), Decimal('59.46'), Decimal('0.77')),
 (40, 'Yuba County', 48, Decimal('14'), Decimal('29.2'), Decimal('59.52'), Decimal('0.83')),
 (41, 'Imperial County', 70, Decimal('22'), Decimal('31.4'), Decimal('59.71'), Decimal('1.02')),
 (42, 'Ventura County', 102

While Mono and Alpine county have significantly lower average churn scores than the rest of the state, this is where having contextual information provides valuable insight into what is really happening. The top two counties have very few customers, as they are sparsely populated, with only a few customers churning. This could be driving the low average churn score they are recieving. I still believe that they should be targeted for marketing, but not as heavily as the difference from the statewide average suggests. 